README.md
# Veeam:
### _a cookbook to deploy Veeam Backup and Replication server_
---
This cookbook installs and configures Veeam Backup and Replication based on documented Veeam best practices.  The cookbook includes recipes to deploy all components of the solution as well as the optional Explorers.

_Note: Veeam prerequisites requires that Microsoft .NET Framework 4.5.2 be installed on the host.  As part of the installation, a reboot is required and will automatically be handled by the resource_

## Table of Contents
*generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Veeam:](#veeam)
  - [### _a cookbook to deploy Veeam Backup and Replication server_](#h3-ida-cookbook-to-deploy-veeam-backup-and-replication-server-34a-cookbook-to-deploy-veeam-backup-and-replication-serverh3)
  - [Table of Contents](#table-of-contents)
  - [Requirements](#requirements)
    - [Platforms](#platforms)
      - [_Note regarding SQL Express Requirements:_](#note-regarding-sql-express-requirements)
    - [Chef](#chef)
    - [Cookbooks](#cookbooks)
  - [Attributes](#attributes)
    - [Installation Media](#installation-media)
    - [Upgrade Details](#upgrade-details)
    - [Catalog](#catalog)
    - [Server](#server)
    - [Console](#console)
    - [Proxy](#proxy)
    - [Host](#host)
  - [Veeam Media and Licenses](#veeam-media-and-licenses)
    - [Veeam Backup and Replication ISO](#veeam-backup-and-replication-iso)
    - [Veeam Backup and Replication Update Zip files](#veeam-backup-and-replication-update-zip-files)
    - [Veeam Backup and Replication License file](#veeam-backup-and-replication-license-file)
  - [Veeam Upgrade Procedures and Details](#veeam-upgrade-procedures-and-details)
    - [Configuring the Updates](#configuring-the-updates)
    - [Upgrade Process and Warnings](#upgrade-process-and-warnings)
    - [Recipes that perform automatic upgrades](#recipes-that-perform-automatic-upgrades)
  - [Resource/Provider](#resourceprovider)
    - [Veeam_Prerequisites](#veeam_prerequisites)
      - [Actions:](#actions)
      - [Properties:](#properties)
      - [Examples:](#examples)
    - [Veeam_Catalog](#veeam_catalog)
      - [Actions:](#actions-1)
      - [Properties:](#properties-1)
      - [Examples:](#examples-1)
    - [Veeam_Console](#veeam_console)
      - [Actions:](#actions-2)
      - [Properties:](#properties-2)
      - [Examples:](#examples-2)
    - [Veeam_Server](#veeam_server)
      - [Actions:](#actions-3)
      - [Properties:](#properties-3)
      - [Examples:](#examples-3)
    - [Veeam_Explorer](#veeam_explorer)
      - [Actions:](#actions-4)
      - [Properties:](#properties-4)
      - [Examples:](#examples-4)
    - [Veeam_Upgrade](#veeam_upgrade)
      - [Actions:](#actions-5)
      - [Properties:](#properties-5)
      - [Examples:](#examples-5)
    - [Veeam_Proxy](#veeam_proxy)
      - [Actions:](#actions-6)
      - [Properties:](#properties-6)
      - [Examples:](#examples-6)
    - [Veeam_Host](#veeam_host)
      - [Actions:](#actions-7)
      - [Properties:](#properties-7)
      - [Examples:](#examples-7)
  - [Usage](#usage)
    - [default](#default)
    - [catalog recipe](#catalog-recipe)
    - [server recipe](#server-recipe)
    - [console recipe](#console-recipe)
    - [server_with_catalog recipe](#server_with_catalog-recipe)
    - [server_with_console recipe](#server_with_console-recipe)
    - [standalone_complete recipe](#standalone_complete-recipe)
    - [proxy recipe](#proxy-recipe)
    - [proxy_remove recipe](#proxy_remove-recipe)
    - [upgrade recipe](#upgrade-recipe)
    - [host_mgmt recipe](#host_mgmt-recipe)
  - [Upload to Chef Server](#upload-to-chef-server)
  - [Matchers/Helpers](#matchershelpers)
    - [Matchers](#matchers)
    - [Veeam::Helper](#veeamhelper)
      - [check_os_version](#check_os_version)
      - [find_package_url(version)](#find_package_urlversion)
      - [find_package_checksum(version)](#find_package_checksumversion)
      - [find_update_url(version)](#find_update_urlversion)
      - [find_update_checksum(version)](#find_update_checksumversion)
      - [prerequisites_list](#prerequisites_list)
      - [explorers_list](#explorers_list)
      - [find_current_dotnet](#find_current_dotnet)
      - [validate_powershell_out(script, timeout: nil)](#validate_powershell_outscript-timeout-nil)
      - [find_current_veeam_solutions(package_name)](#find_current_veeam_solutionspackage_name)
      - [find_current_veeam_version(package_name)](#find_current_veeam_versionpackage_name)
      - [iso_installer(downloaded_file_name, new_resource)](#iso_installerdownloaded_file_name-new_resource)
      - [extract_installer(downloaded_file_name, new_resource)](#extract_installerdownloaded_file_name-new_resource)
      - [unmount_installer(downloaded_file_name)](#unmount_installerdownloaded_file_name)
      - [get_media_installer_location(downloaded_file_name)](#get_media_installer_locationdownloaded_file_name)
    - [Windows_Helper](#windows_helper)
  - [Cookbook Testing](#cookbook-testing)
    - [Before you begin](#before-you-begin)
    - [Data_bags for Test-Kitchen](#data_bags-for-test-kitchen)
    - [Rakefile and Tasks](#rakefile-and-tasks)
    - [Chefspec and Test-Kitchen](#chefspec-and-test-kitchen)
    - [Compliance Profile](#compliance-profile)
  - [Contribute](#contribute)
  - [License and Author](#license-and-author)

## Requirements

### Platforms
- Windows Server 2012
- Windows Server 2012R2
- Windows Server 2016

Windows 2008R2 and lower is _not_ supported.

#### _Note regarding SQL Express Requirements:_
The installation of SQL Express requires that a temporary Scheduled Task be created within Windows to perform the installation.  The reason for this workaround is due to a known limitation with Microsoft SQL installations via remote powershell executions.  This enhancement has been added to allow for the ability to perform the installation via Terraform, Knife Bootstrap, or any other remote powershell based execution process.

### Chef

- Chef 12.5+

### Cookbooks

- windows


## Attributes
### Installation Media
| Attribute | Type | Description | Default Value | Mandatory |
| --- | --- | --- | --- | --- |
| `node['veeam']['version']` | String. | Base version of Veeam to install and used to download the appropriate ISO.  Supported versions are '9.0' and '9.5' | '9.5' | |
| `node['veeam']['installer']['package_url']` | String. | Custom URL for the Veeam Backup and Replication ISO. If not provided, then the ISO will be downloaded directly from Veeam | nil | |
| `node['veeam']['installer']['package_checksum']` | String. | Sha256 hash of the remote ISO file. Required when setting the `node['veeam']['installer']['package_url']`| nil | |
| `node['veeam']['license_url']` | String. | URL for downloading the license filed used by this server.  If not provided, the [license data_bag](#veeam-backup-and-replication-license-file) will be checked or the software will be installed in evaluation mode. | nil | |

### Upgrade Details
| Attribute | Type | Description | Default Value | Mandatory |
| --- | --- | --- | --- | --- |
| `node['veeam']['installer']['update_url']` | String. | Custom URL for the Veeam Backup and Replication ISO or Upgrade ZIP. If not provided, then the `node['veeam']['installer']['package_url']` will be used.  | nil | |
| `node['veeam']['installer']['update_checksum']` | String. | Sha256 hash of the remote ISO or ZIP file.  Required when setting the `node['veeam']['installer']['update_url']` | nil |
| `node['veeam']['build']` | String | Current Veeam Build ID to be used when performing upgrades.  Will default to the value of the build found in the attribute `node['veeam']['installer']['update_url']` unless not set.  If no `node['veeam']['installer']['update_url']` provided then the value of the Build in `node['veeam']['installer']['package_url']` will be used.  Otherwise, the Value of the `node['veeam']['version']` will be assigned. | Multi | |
| `node['veeam']['reboot_on_upgrade']` | TrueFalse | When performing an upgrade, the Veeam process will sometimes require a reboot.  This key will control if the reboot should be automatically performed at the end of the upgrade. | true | |
| `node['veeam']['upgrade']['keep_media']` | TrueFalse | Determines if the recipe should keep the media at the end of the upgrade. | false | |

### Catalog
| Attribute | Type | Description | Default Value | Mandatory |
| --- | --- | --- | --- | --- |
| `node['veeam']['catalog']['install_dir']` | String. | Installs the component to the specified location. By default, Veeam Backup & Replication uses the Backup Catalog subfolder in the `C:\Program Files\Veeam\Backup and Replication\` folder. | C:\Program Files\Veeam\Backup and Replication\ | |
| `node['veeam']['catalog']['vm_catalogpath']` | String. |  Specifies a path to the catalog folder where index files must be stored. By default, Veeam Backup & Replication uses the C:\VBRCatalog folder to store index files. | C:\VBRCatalog | |
| `node['veeam']['catalog']['vbrc_service_user']` | String. | Specifies a user account under which the Veeam Guest Catalog Service will run. The account must have full control NTFS permissions on the `VM_CATALOGPATH` folder where index files are stored.  If you do not specify this parameter, the Veeam Guest Catalog Service will run under the Local System account.  NOTE: The account must be in Domain\User or Computer\User format.  If using a local account, then use either the `hostname\username` or use `.\username` | nil | |
| `node['veeam']['catalog']['vbrc_service_password']` | String. | Specifies a password for the account under which the Veeam Guest Catalog Service will run.  This parameter must be used if you have specified the `VBRC_SERVICE_USER` parameter. | nil | |
| `node['veeam']['catalog']['vbrc_service_port']` | Integer. | Specifies a TCP port that will be used by the Veeam Guest Catalog Service. By default, port number 9393 is used. | 9393 | |
| `node['veeam']['catalog']['keep_media']` | TrueFalse. | Determines if the recipe should keep the media at the end of the installation. | false | |

### Server
| Attribute | Type | Description | Default Value | Mandatory |
| --- | --- | --- | --- | --- |
| `node['veeam']['server']['accept_eula']` | TrueFalse |  Must be set to true or the server will not install.  Since we can download the media directly, it is a good idea to enforce the EULA. | false | X |
| `node['veeam']['server']['install_dir']` | String | Installs the component to the specified location. By default, Veeam Backup & Replication uses the Backup Server subfolder in the `C:\Program Files\Veeam\Backup and Replication\` folder. | C:\Program Files\Veeam\Backup and Replication\ | |
| `node['veeam']['server']['evaluation']` | TrueFalse | Determines if the Veeam Backup and Replication server should be installed using Evaluation Mode or if a license should be attached.  Default value is true and the server will be installed with no license. | true | |
| `node['veeam']['server']['vbr_check_updates']` | TrueFalse | Specifies if you want Veeam Backup & Replication to automatically check for new product patches and versions. | false | |
| `node['veeam']['server']['vbr_service_user']` | String | Specifies the account under which the Veeam Backup Service will run. The account must have full control NTFS permissions on the `VBRCatalog` folder where index files are stored and the Database owner rights for the configuration database on the Microsoft SQL Server where the configuration database is deployed.  If you do not specify this parameter, the Veeam Guest Catalog Service will run under the Local System account.  NOTE: The account must be in Domain\User or Computer\User format.  If using a local account, then use either the `hostname\username` or use `.\username` | Local System | |
| `node['veeam']['server']['vbr_service_password']` | String | Specifies a password for the account under which the Veeam Guest Backup Service will run.  This parameter must be used if you have specified the `VBR_SERVICE_USER` parameter. | nil | |
| `node['veeam']['server']['vbr_service_port']` | Integer | Specifies a TCP port that will be used by the Veeam Guest Backup Service. By default, port number 9392 is used. | 9392 | |
| `node['veeam']['server']['vbr_secure_connections_port']` | Integer | Specifies an SSL port used for communication between the mount server and the backup server. By default, port 9401 is used. | 9401 | |
| `node['veeam']['server']['vbr_sqlserver_server']` | String | Specifies a Microsoft SQL server and instance on which the configuration database will be deployed. By default, Veeam Backup & Replication uses the (local)\VEEAMSQL2012 server.  If not included or set, the recipe will install SQLExpress 2012 on the node. | nil | |
| `node['veeam']['server']['vbr_sqlserver_database']` | String | Specifies a name of the configuration database to be deployed. | VeeamBackup | |
| `node['veeam']['server']['vbr_sqlserver_auth']` | String | Specifies if you want to use the SQL Server authentication mode to connect to the Microsoft SQL Server where the Veeam Backup & Replication is deployed.  Supported Values are Windows or Mixed | nil | |
| `node['veeam']['server']['vbr_sqlserver_username']` | String | This parameter must be used if you have specified the `VBR_SQLSERVER_AUTHENTICATION` parameter.  Specifies a LoginID to connect to the Microsoft SQL Server in the SQL Server authentication mode. | nil | |
| `node['veeam']['server']['vbr_sqlserver_password']` | String | This parameter must be used if you have specified the `VBR_SQLSERVER_AUTHENTICATION` parameter.  Specifies a password to connect to the Microsoft SQL Server in the SQL Server authentication mode. | nil | |
| `node['veeam']['server']['pf_ad_nfsdatastore']` | String | Specifies the vPower NFS root folder to which Instant VM Recovery cache will be stored. | C:\ProgramData\Veeam\Backup\NfsDatastore\ | |
| `node['veeam']['server']['keep_media']` | TrueFalse |  Determines if the recipe should keep the media at the end of the installation. | false | |
| `node['sql_server']['server_sa_password']` | String | Configures the SQL Admin password for the SQLExpress instance. | 'Veeam1234' | |
| `node['veeam']['server']['explorers']` | Array. List of Veeam Explorers to install. | 'ActiveDirectory','Exchange','SQL','Oracle','SharePoint' | |

### Console
| Attribute | Type | Description | Default Value | Mandatory |
| --- | --- | --- | --- | --- |
| `node['veeam']['console']['accept_eula']` | TrueFalse | Must be set to true or the server will not install.  Since we can download the media directly, it is a good idea to enforce the EULA. | false | X |
| `node['veeam']['console']['install_dir']` | String | Installs the component to the specified location. By default, Veeam Backup & Replication uses the Backup Console subfolder in the `C:\Program Files\Veeam\Backup and Replication\` folder. | C:\Program Files\Veeam\Backup and Replication\ | |
| `node['veeam']['console']['keep_media']` | TrueFalse | Determines if the recipe should keep the media at the end of the installation. | false | |

### Proxy
| Attribute | Type | Description | Default Value | Mandatory |
| --- | --- | --- | --- | --- |
| `node['veeam']['proxy']['vbr_server']` | String | DNS or IP Address of the Veeam Backup and Replication Server |  | X |
| `node['veeam']['proxy']['vbr_port']` | String | Veeam Backup and Replication Server Port | 9392 | |
| `node['veeam']['proxy']['vbr_username']` | String | Username with permissions to add Servers and Proxies within Veeam Backup and Replication server |  | X |
| `node['veeam']['proxy']['vbr_password']` | String | Password for the user provided |  | X |
| `node['veeam']['proxy']['proxy_username']` | String | Required when Adding a new Proxy.  Username with Access to the Proxy Server.  Will generate a new set of credentials within Veeam Backup and Replication server if none exist.|  | X |
| `node['veeam']['proxy']['proxy_password']` | String | Required when Adding a new Proxy.  Password for the user provided.|  | = nil
| `node['veeam']['proxy']['description']` | String | Description for the Proxy Server.  Will automatically start with "ADDED by CHEF: ".  Defaults to "ADDED by CHEF: Proxy Server" |  | X |

CHANGELOG.md
# Change log information for Veeam Cookbook

## Version 3.0.6
2020-07-31

Reboot loop bug fix for update 4b (9.5.4.2866) that has different desired version for vbr component.

## Version 3.0.3
2020-07-30

Support update 4b (9.5.4.2866)

## Version 3.0.2
2020-07-24

Minor fix update for SQL Express installation on domain joined computer.

## Version 3.0.1
2020-07-14

Minor fix update to include `delay_min 1` to all reboot resources to cover a chef bug that lets some test-kitchen environments crash.

## Version 2.1.1
2018-08-25

Minor fix update to include `auto_reboot` override to the Recipe::Proxy and Recipe::HostMgmt

## Version 2.1
2018-08-24

Minor update to include default links for downloading Update 3a.  Also added a new resource and recipe to allow for automatic adding of a host to the Server Inventory of Veeam Backup and Replication.

### Included:
- Updated Helpers to include support for Veeam 9.5.0.1922 (Update 3a)
- Add resource to control adding new Hosts into Veeam Inventory.

### Information on new resources:

* **VeeamHost** - Registers the provided server as the selected type by connecting to the Veeam Backup and Replication server via the Veeam PowerShell toolkit.  This resource will add a Veeam credential object if one does not exist and then register the server.

## Version 2.0
2018-05-04

Major update to the Veeam cookbook with full testing against Chef client 14.0.202.  This update includes two new custom resources - VeeamProxy and VeeamUpgrade. Along with the new resources, a complete refactor of existing resources was completed to consolidate the helper methods into a single library.

### Included:
- Veeam Backup and Replication ProxyServer deployments
- Support for automatic Veeam Backup and Replication component Updates
- Refactor of existing Custom Resources and Helper Library

### Information on new resources:

* **VeeamProxy** - Configures a Windows server as either a VMware or HyperV Backup Proxy by connecting to the Veeam Backup and Replication server via the Veeam PowerShell toolkit.  This resource will add a Veeam credential object if one does not exist and then register the server as a Proxy Type.
* **VeeamUpgrade** - The process to perform upgrades requires that the appropriate installation media is provided which contains the updates from Veeam.  This cookbook will initiate an upgrade if the currently installed versions are less than the desired Build version as defined by the attribute `node['veeam']['build']`.  When the installed version does not match the requested build version, the process will mount the ISO or extract the ZIP that contains the update and then perform an automatic upgrade of each service installed on the host.

### Major Updates:
- Veeam::Helper refactor
- SQL Express installation via Windows Scheduled task (resolves issues with remote builds on domain joined servers)


This cookbook is not officially supported by Veeam Software

## Version 1.0
2017-11-05

Initial release of the Veeam Cookbook including support for the new provisioning

### Included:
- Veeam Backup and Replication Server deployments
- Veeam Backup and Replication Catalog deployments
- Veeam Backup and Replication Console deployments
- Veeam Backup and Replication Prerequisite deployments
- Custom Resources and Helper Library


### Information on new resources:

* **VeeamPrerequisites** - Deploys all Veeam Prerequisite packages including SQL Tools, .NET 4.5, and SQL Express.
* **VeeamConsole** - Installs the Veeam Backup and Replication Console
* **VeeamCatalog** - Installs the Veeam Backup and Replication Catalog
* **VeeamServer** - Installs the Veeam Backup and Replication Server
* **VeeamExplorers** - Installs the Veeam Backup and Replication Explorers based on those available in the Installation Media.

This cookbook is not officially supported by Veeam Software


libraries/helper.rb
# Cookbook Name:: veeam
# Library:: helper
#
# Author:: Jeremy Goodrum
# Email:: chef@exospheredata.com
#
# Version:: 0.1.0
# Date:: 2017-02-07
#
# Copyright (c) 2016 Exosphere Data LLC, All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'chef/mixin/shell_out'
require 'chef/util/path_helper'

module Veeam
  module Helper
    def check_os_version(node)
      # '6.1.' is the numeric platform_version for Windows 2008R2.  If the node OS version is below that value, we must abort.
      raise ArgumentError, 'Veeam Backup and Replication management requires a Windows 2008R2 or higher host!' if node['platform_version'].to_f < '6.1'.to_f
      # If the kernel is not 64bit then raise an error, as we cannot proceed.
      raise ArgumentError, 'Veeam Backup and Replication requires an x86_64 host and cannot be installed on this machine' unless node['kernel']['machine'] =~ /x86_64/
    end

    def find_package_url(version)
      package_list(version)['package_url'] if package_list(version)
    end

    def find_package_checksum(version)
      package_list(version)['package_checksum'] if package_list(version)
    end

    def package_list(version)
      case version.to_s # to_s to make sure someone didn't pass us an int
      when '9.0' then {
        'package_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.0.0.902.iso',
        'package_checksum' => '21f9d2c318911e668511990b8bbd2800141a7764cc97a8b78d4c2200c1225c88'
      }
      when '9.5' then {
        'package_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.711.iso',
        'package_checksum' => 'af3e3f6db9cb4a711256443894e6fb56da35d48c0b2c32d051960c52c5bc2f00'
      }
      when '9.5.0.711' then {
        'package_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.711.iso',
        'package_checksum' => 'af3e3f6db9cb4a711256443894e6fb56da35d48c0b2c32d051960c52c5bc2f00'
      }
      when '9.5.0.1038' then {
        'package_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.1038.Update2.iso',
        'package_checksum' => '180b142c1092c89001ba840fc97158cc9d3a37d6c7b25c93a311115b33454977'
      }
      when '9.5.0.1536' then {
        'package_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.1536.Update3.iso',
        'package_checksum' => '5020ef015e4d9ff7070d43cf477511a2b562d8044975552fd08f82bdcf556a43'
      }
      when '9.5.0.1922' then {
        'package_url' => 'https://download2.veeam.com/VeeamBackup&Replication_9.5.0.1922.Update3a.iso',
        'package_checksum' => '9a6fa7d857396c058b2e65f20968de56f96bc293e0e8fd9f1a848c7d71534134'
      }
      when '9.5.4.2615' then {
        'package_url' => 'https://download2.veeam.com/VeeamBackup&Replication_9.5.4.2615.Update4.iso',
        'package_checksum' => 'ecc27bbcf49104861566782701dca42375b324b4710e2fa79b5f8068c31c4494'
      }
      when '9.5.4.2866' then {
        'package_url' => 'https://download2.veeam.com/VeeamBackup&Replication_9.5.4.2866.Update4b_20191210.iso',
        'package_checksum' => 'cfc41596154563f60b74320634589721fd1110c87e04632068bc5234aada342e'
      }
    end
    end

    def find_update_url(version)
      update_list(version)['update_url'] if update_list(version)
    end

    def find_update_checksum(version)
      update_list(version)['update_checksum'] if update_list(version)
    end

    def update_list(version)
      case version.to_s # to_s to make sure someone didn't pass us an int
      when '9.0' then {
        'update_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.0.0.902.iso',
        'update_checksum' => '21f9d2c318911e668511990b8bbd2800141a7764cc97a8b78d4c2200c1225c88'
      }
      when '9.5' then {
        'update_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.711.iso',
        'update_checksum' => 'af3e3f6db9cb4a711256443894e6fb56da35d48c0b2c32d051960c52c5bc2f00'
      }
      when '9.5.0.711' then {
        'update_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.711.iso',
        'update_checksum' => 'af3e3f6db9cb4a711256443894e6fb56da35d48c0b2c32d051960c52c5bc2f00'
      }
      when '9.5.0.823' then {
        'update_url' => 'https://download2.veeam.com/VeeamBackup&Replication_9.5.0.823_Update1.zip',
        'update_checksum' => 'c07bdfb3b90cc609d21ba94584ba19d8eaba16faa31f74ad80814ec9288df492'
      }
      when '9.5.0.1038' then {
        'update_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.1038.Update2.zip',
        'update_checksum' => 'd800bf5414f1bde95fba5fddbd86146c75a5a2414b967404792cc32841cb4ffb'
      }
      when '9.5.0.1536' then {
        'update_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.1536.Update3.zip',
        'update_checksum' => '38ed6a30aa271989477684fdfe7b98895affc19df7e1272ee646bb50a059addc'
      }
      when '9.5.0.1922' then {
        'update_url' => 'http://download2.veeam.com/VeeamBackup&Replication_9.5.0.1922.Update3a.zip',
        'update_checksum' => 'f6b3fc0963b09362c535ef49691c51d368266cc91d6833c80c70342161bb7123'
      }
      when '9.5.4.2615' then {
        'package_url' => 'https://download2.veeam.com/VeeamBackup&Replication_9.5.4.2615.Update4.iso',
        'package_checksum' => 'ecc27bbcf49104861566782701dca42375b324b4710e2fa79b5f8068c31c4494'
      }
      when '9.5.4.2866' then {
        'package_url' => 'https://download2.veeam.com/VeeamBackup&Replication_9.5.4.2866.Update4b_20191210.iso',
        'package_checksum' => 'cfc41596154563f60b74320634589721fd1110c87e04632068bc5234aada342e'
      }
      end
    end

    def prerequisites_list(version)
      case version.to_s # to_s to make sure someone didn't pass us an int
      when '9.0' then {
        '0' => { 'Microsoft System CLR Types for SQL Server 2012 (x64)' => 'SQLSysClrTypes.msi' },
        '1' => { 'Microsoft SQL Server 2012 Management Objects  (x64)' => 'SharedManagementObjects.msi' },
        'SQL' => { 'Microsoft SQL Server 2012 (64-bit)' => 'SQLEXPR_x64_ENU.exe' }
      }
      when /9.0.0.\d+/ then {
        '0' => { 'Microsoft System CLR Types for SQL Server 2012 (x64)' => 'SQLSysClrTypes.msi' },
        '1' => { 'Microsoft SQL Server 2012 Management Objects  (x64)' => 'SharedManagementObjects.msi' },
        'SQL' => { 'Microsoft SQL Server 2012 (64-bit)' => 'SQLEXPR_x64_ENU.exe' }
      }
      when '9.5' then {
        '0' => { 'Microsoft System CLR Types for SQL Server 2014' => 'SQLSysClrTypes.msi' },
        '1' => { 'Microsoft SQL Server 2014 Management Objects  (x64)' => 'SharedManagementObjects.msi' },
        'SQL' => { 'Microsoft SQL Server 2012 (64-bit)' => 'SQLEXPR_x64_ENU.exe' }
      }
      when /9.5.0.\d+/ then {
        '0' => { 'Microsoft System CLR Types for SQL Server 2014' => 'SQLSysClrTypes.msi' },
        '1' => { 'Microsoft SQL Server 2014 Management Objects  (x64)' => 'SharedManagementObjects.msi' },
        'SQL' => { 'Microsoft SQL Server 2012 (64-bit)' => 'SQLEXPR_x64_ENU.exe' }
      }
      when /9.5.4.\d+/ then {
        '0' => { 'Microsoft System CLR Types for SQL Server 2014' => 'SQLSysClrTypes.msi' },
        '1' => { 'Microsoft SQL Server 2014 Management Objects  (x64)' => 'SharedManagementObjects.msi' },
        'SQL' => { 'Microsoft SQL Server 2016 (64-bit)' => 'SQLEXPR_x64_ENU.exe' }
      }
      end
    end

    def explorers_list(version)
      case version.to_s # to_s to make sure someone didn't pass us an int
      when '9.0' then {
        'ActiveDirectory' => 'Veeam Explorer for Microsoft Active Directory',
        'SQL' => 'Veeam Explorer for Microsoft SQL Server',
        'Exchange' => 'Veeam Explorer for Microsoft Exchange',
        'SharePoint' => 'Veeam Explorer for Microsoft SharePoint',
        'Oracle' => 'Veeam Explorer for Oracle'
      }
      when '9.5' then {
        'ActiveDirectory' => { name: 'Veeam Explorer for Microsoft Active Directory', version: '9.5.0.836' },
        'SQL' => { name: 'Veeam Explorer for Microsoft SQL Server', version: '9.5.0.836' },
        'Exchange' => { name: 'Veeam Explorer for Microsoft Exchange', version: '9.5.0.836' },
        'SharePoint' => { name: 'Veeam Explorer for Microsoft SharePoint', version: '9.5.0.836' },
        'Oracle' => { name: 'Veeam Explorer for Oracle', version: '9.5.0.836' }
      }
      when /9.5.0.\d+/ then {
        'ActiveDirectory' => { name: 'Veeam Explorer for Microsoft Active Directory', version: '9.5.0.836' },
        'SQL' => { name: 'Veeam Explorer for Microsoft SQL Server', version: '9.5.0.836' },
        'Exchange' => { name: 'Veeam Explorer for Microsoft Exchange', version: '9.5.0.836' },
        'SharePoint' => { name: 'Veeam Explorer for Microsoft SharePoint', version: '9.5.0.836' },
        'Oracle' => { name: 'Veeam Explorer for Oracle', version: '9.5.0.836' }
      }
      when '9.5.4.2615' then {
        'ActiveDirectory' => { name: 'Veeam Explorer for Microsoft Active Directory', version: '9.6.4.1053' },
        'SQL' => { name: 'Veeam Explorer for Microsoft SQL Server', version: '9.6.4.1053' },
        'Exchange' => { name: 'Veeam Explorer for Microsoft Exchange', version: '9.6.4.1053' },
        'SharePoint' => { name: 'Veeam Explorer for Microsoft SharePoint', version: '9.6.4.1053' },
        'Oracle' => { name: 'Veeam Explorer for Oracle', version: '9.6.4.1053' }
      }
      when '9.5.4.2866' then {
        'ActiveDirectory' => { name: 'Veeam Explorer for Microsoft Active Directory', version: '9.6.4.1090' },
        'SQL' => { name: 'Veeam Explorer for Microsoft SQL Server', version: '9.6.4.1092' },
        'Exchange' => { name: 'Veeam Explorer for Microsoft Exchange', version: '9.6.4.1090' },
        'SharePoint' => { name: 'Veeam Explorer for Microsoft SharePoint', version: '9.6.4.1090' },
        'Oracle' => { name: 'Veeam Explorer for Oracle', version: '9.6.4.1090' }
      }
      end
    end

    def find_current_dotnet
      installed_version = nil
      installed_version_reg_key = registry_get_values('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full')
      unless installed_version_reg_key.nil?
        installed_version_reg_key.each do |key|
          installed_version = key[:data] if key[:name] == 'Release'
        end
      end
      installed_version.nil? ? 0 : installed_version
    end

    def validate_powershell_out(script, timeout: nil, ignore_errors: false)
      # This seemed like the DRYest way to handle the output handling from PowerShell.
      cmd = powershell_out(script) if timeout.nil?
      cmd = powershell_out(script, timeout: timeout) unless timeout.nil?
      # Only return the output if there were no errors.
      return cmd.stdout.chomp if cmd.stderr == '' || cmd.stderr.nil? || ignore_errors
      raise cmd.inspect if cmd.stderr != ''
    end

    def find_current_veeam_solutions(package_name)
      cmd_str = <<-EOH
        $program = '#{package_name}'
        $x86 = (Get-ChildItem "HKLM:\\Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall" | gp )
        $x64 = (Get-ChildItem "HKLM:\\Software\\Wow6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall" | gp )

        $x86version = ($x86 | ?{$_.DisplayName -eq $program})
        $x64version = ($x64 | ?{$_.DisplayName -eq $program})

        if($x86version) {
          return ($x86version.InstallLocation)
        } elseif($x64version) {
          return ($x64version.InstallLocation)
        } else {
          # Return nothing
        }
      EOH
      output = validate_powershell_out(cmd_str)
      raise ArgumentError, 'Unable to find the Veeam installation' unless output
      output
    end

    def find_current_veeam_version(package_name)
      veeam_package = find_current_veeam_solutions(package_name)
      veeam_exe = case package_name
                  when /Console/
                    "#{veeam_package}\\Console\\veeam.backup.shell.exe"
                  when /Server/
                    "#{veeam_package}\\Backup\\Packages\\VeeamDeploymentDll.dll"
                  when /Catalog/
                    "#{veeam_package}\\Backup Catalog\\VeeamDeploymentDll.dll"
                  else
                    raise "Unknown Package name: #{package_name}"
                  end
      cmd_str = <<-EOH
        $File = Get-Item -Path '#{veeam_exe}'
        $File.VersionInfo.ProductVersion
      EOH
      output = validate_powershell_out(cmd_str, ignore_errors: true)
      output = nil if output == ''
      output
    end

    def iso_installer(downloaded_file_name, new_resource)
      remote_file downloaded_file_name do
        source new_resource.package_url
        checksum new_resource.package_checksum
        use_conditional_get true # this should allow us to prevent duplicate downloads
        action :create
      end

      # Mounting the Veeam backup ISO.
      powershell_script 'Load Veeam media' do
        code <<-EOH
          Mount-DiskImage -ImagePath "#{downloaded_file_name}"
        EOH
        action :run
        guard_interpreter :powershell_script
        not_if "[boolean] (Get-DiskImage -ImagePath '#{downloaded_file_name}').DevicePath"
      end
    end

    def extract_installer(downloaded_file_name, new_resource)
      package_name = downloaded_file_name.split('/').last
      package_type = ::File.extname(package_name)
      install_media_path = win_clean_path(::File.join(::Chef::Config[:file_cache_path], "Veeam/#{package_name.gsub(package_type, '')}"))
      update_path = win_clean_path(::File.join(install_media_path, '/Updates'))

      remote_file downloaded_file_name do
        source new_resource.package_url
        checksum new_resource.package_checksum
        use_conditional_get true # this should allow us to prevent duplicate downloads
        action :create
        not_if { ::File.exist?(update_path) }
      end

      windows_zipfile win_clean_path(::File.join(install_media_path, '/Updates')) do
        source downloaded_file_name
        action :unzip
        not_if { ::File.exist?(update_path) }
      end

      install_media_path
    end

    def unmount_installer(downloaded_file_name)
      # Unmount the Veeam backup ISO.
      powershell_script 'Dismount Veeam media' do
        code <<-EOH
          Dismount-DiskImage -ImagePath "#{downloaded_file_name}"
        EOH
        action :run
        guard_interpreter :powershell_script
        only_if "[boolean] (Get-DiskImage -ImagePath '#{downloaded_file_name}').DevicePath"
      end
    end

    def get_media_installer_location(downloaded_file_name)
      # When downloading and mounting the ISO, we need to track back to the Drive Letter.  This method will handle
      # the look-up and keep the logic out of the main installation code.
      Chef::Log.debug 'Searching for the Veeam installation media Drive Letter...'
      cmd_str = <<-EOH
        $DriveLetter = (Get-DiskImage -ImagePath '#{downloaded_file_name}' | Get-Volume).DriveLetter;
        if ( [string]::IsNullOrEmpty($DriveLetter) ){ throw 'The ISO did not mount and we have no idea why.' }
        return ( $DriveLetter +':\' )
      EOH
      output = validate_powershell_out(cmd_str)
      raise ArgumentError, 'Unable to find the Veeam installation media' unless output
      Chef::Log.debug "Found the Veeam installation media at Drive Letter [#{output}]"
      output
    end

    def win_clean_path(path)
      Chef::Util::PathHelper.cleanpath(path)
    end
  end
end
